### PR TITLE
[Search MSI 9] Add support for managed identities and delete classic similarity hack

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/Engineering/issues/4103.
Spec: https://github.com/NuGet/Engineering/blob/main/Server.Specs/AzureSearchManagedIdentities.md

Per https://github.com/NuGet/Engineering/pull/4268, we will test BM25 immediately so the classic similarity hack can be deleted. Also, update the Azure Search SDK package version so that we start using managed identities.

This has been verified end-to-end in DEV environment. 

Also, fix a null tags bug found when running Catalog2AzureSearch in DEV (caused by the new SDK).

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/AzureSearchConfiguration.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@ public class AzureSearchConfiguration
     {
         public string SearchServiceName { get; set; }
         public string SearchServiceApiKey { get; set; }
+        public string SearchServiceManagedIdentityClientId { get; set; }
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/AzureSearchConfiguration.cs:12]: It took me a bit to realize this is for a user-assigned managed identity. Consider adding a quick comment to indicate that

**Review by loic-sharma — APPROVED** [review]
Yay great work! 🚀

File: src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs [hunk 2]
```diff
@@ -249,17 +250,27 @@ private static void RegisterAuxiliaryDataStorageServices(ContainerBuilder contai
...
+                        return new SearchIndexClient(
+                            endpoint,
+                            new ManagedIdentityCredential(options.Value.SearchServiceManagedIdentityClientId),
+                            searchOptions);
+                    }
+                    else
+                    {
+                        return new SearchIndexClient(
+                            endpoint,
+                            new AzureKeyCredential(options.Value.SearchServiceApiKey),
```
Comment (shishirx34) [inline src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs:271]: perhaps for coverage also add the check for `SearchServiceApiKey` being not null? Is there any place where we are validating that either the `ManagedIdentityClientId` or `ApiKey` is provided in the configs i.e. either one is a Required config?

**Review by shishirx34 — APPROVED** [review]

File: src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/DependencyInjectionExtensions.cs:280]: Done.

**Review by shishirx34 — APPROVED** [review]

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs [hunk 3]
```diff
@@ -165,7 +165,7 @@ public BaseDocumentBuilder(IOptionsSnapshot<AzureSearchJobConfiguration> options
             document.SemVerLevel = leaf.IsSemVer2() ? SemVerLevelKey.SemVer2 : SemVerLevelKey.Unknown;
             document.SortableTitle = GetSortableTitle(leaf.Title, leaf.PackageId);
             document.Summary = leaf.Summary;
-            document.Tags = leaf.Tags == null ? null : leaf.Tags.ToArray();
+            document.Tags = leaf.Tags == null ? Array.Empty<string>() : leaf.Tags.ToArray();
```
Comment (agr) [inline src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs:168]: Is this change related to authentication?

**Review by agr — APPROVED** [review]

File: src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/BaseDocumentBuilder.cs:168]: Nope, it was found when testing Catalog2AzureSearch in DEV.
